### PR TITLE
8288425: Footprint regression due MH creation when initializing StringConcatFactory

### PR DESCRIPTION
**Pull request by cl4es**
Avoid doing MH creation when initializing `StringConcatFactory` by lazily initializing to a `@Stable` field instead.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8288425](https://bugs.openjdk.org/browse/JDK-8288425): Footprint regression due MH creation when initializing StringConcatFactory

### Reviewers
 * [Jorn Vernee](https://openjdk.java.net/census#jvernee) (@JornVernee - **Reviewer**)
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9154/head:pull/9154` \
`$ git checkout pull/9154`

Update a local copy of the PR: \
`$ git checkout pull/9154` \
`$ git pull https://git.openjdk.org/jdk pull/9154/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9154`

View PR using the GUI difftool: \
`$ git pr show -t 9154`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9154.diff">https://git.openjdk.org/jdk/pull/9154.diff</a>

</details>

### CODE REVIEW

**Review by JornVernee — APPROVED** [review]

**Review by mlchung — APPROVED** [review]

**Comment by cl4es** [comment]
Thanks for reviewing!

/integrate

File: src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java [hunk 1]
```diff
@@ -722,8 +722,15 @@ private static int classIndex(Class<?> cl) {
     private static final int[] PREPEND_FILTER_SECOND_PAIR_ARGS = new int[] { 0, 1, 4, 5 };
 
     // Base MH for complex prepender combinators.
-    private static final MethodHandle PREPEND_BASE = MethodHandles.dropArguments(
-            MethodHandles.identity(long.class), 1, byte[].class);
+    private static @Stable MethodHandle PREPEND_BASE;
+    private static MethodHandle prependBase() {
+        MethodHandle base = PREPEND_BASE;
+        if (base == null) {
+            base = PREPEND_BASE = MethodHandles.dropArguments(
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java:729]: I'm curious about this idiom - couldn't this result in the `PREPEND_BASE` field being assigned twice, with objects that are morally identical, but could have different identities?

Comment (PaulSandoz) [reply src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java:729]: It's not a violation, it's relying on behavior not discussed by the `@Stable` doc, which refers to such behavior as "undefined".

It's a carefully correlated dance between the code, interpreter, and C2 relying on:
1. safe publication of equivalent immutable structures with no dependence on their identity when operating on them;
2. the field being set in the interpreter; and 
3. by the time C2 inlines `prependBase` the field is no longer be updated, thus C2 can replace the field access with its value.

Comment (mcimadamore) [reply src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java:729]: Thanks, I think I was missing (2)

**Pull request closed without merging**